### PR TITLE
Make no_backprop_mode

### PR DESCRIPTION
**Pull request by unnonouno**
I made an evaluation mode which doesn't make a computational graph for `Variable` whose volatile flag is `AUTO`.

### CODE REVIEW

**Comment by beam2d** [comment]
I think the name `evaluation_mode` might be a bit confusing, because there would be a case that this mode may be used in training as well (e.g. when the loss function includes a term that does not require backprop). I prefer more direct naming, like `no_backprop_mode`.

**Comment by unnonouno** [comment]
It sounds good. Thank you.

File: chainer/function.py [hunk 1]
```diff
@@ -11,6 +13,35 @@
 from chainer import variable
 
 
+_thread_local = threading.local()
+_thread_local.default_backprop = True
+
+
+@contextlib.contextmanager
+def no_backprop_mode():
+    """Disable back-propagation for Variable whose volale is auto.
```
Comment (beam2d) [inline chainer/function.py]: volale -> volatile

File: chainer/function.py [hunk 2]
```diff
@@ -11,6 +13,35 @@
...
+
+    In the default setting :class:`~chainer.Variable` whose `volatile`
```
Comment (beam2d) [inline chainer/function.py]: Variable -> a Variable object

File: chainer/function.py [hunk 3]
```diff
@@ -11,6 +13,35 @@
...
+    attribute is `'auto'` behaves like **non-volatile** variable. That means
```
Comment (beam2d) [inline chainer/function.py]: a non-volatile variable

File: chainer/function.py [hunk 4]
```diff
@@ -11,6 +13,35 @@
...
+    such a :class:`~chainer.Variable`: builds a computational graph, consumes
```
Comment (beam2d) [inline chainer/function.py]: a Variable object

File: chainer/function.py [hunk 5]
```diff
@@ -11,6 +13,35 @@
...
+    memory to store the graph, and you can execute back-propagation for it.
+    With this context such a :class:`~chainer.Variable`: behaves **volatile**
```
Comment (beam2d) [inline chainer/function.py]: a Variable object

Comment (beam2d) [inline chainer/function.py]: behaves volatile variable -> behaves like a volatile variable

File: chainer/function.py [hunk 6]
```diff
@@ -11,6 +13,35 @@
...
+    variable. So, you can easily switch training and evaluation.
```
Comment (beam2d) [inline chainer/function.py]: So -> Then

File: chainer/function.py [hunk 7]
```diff
@@ -11,6 +13,35 @@
...
+
+    In this example volatily of ``x`` and ``y`` is ``'auto'``. So, ``y``
```
Comment (beam2d) [inline chainer/function.py]: In this example -> In this example,

Comment (beam2d) [inline chainer/function.py]: volatily of -> the volatility

File: chainer/function.py [hunk 8]
```diff
@@ -11,6 +13,35 @@
...
+    doesn't have a computational graph.
```
Comment (beam2d) [inline chainer/function.py]: doesn't -> does not

**Comment by beam2d** [comment]
Don't we need "force backprop mode"?

**Comment by unnonouno** [comment]
You mean, all variables behave like volatile='on' in this mode?

**Comment by beam2d** [comment]
I meant the opposite: all 'auto' variables behave like volatile='off' in this mode, which is meaningful only when the code is in no-backprop mode (i.e. I meant the case of nested 'no-backprop' and 'backprop' modes).

**Comment by unnonouno** [comment]
i see

**Comment by unnonouno** [comment]
I fixed it.

**Comment by beam2d** [comment]
LGTM!
